### ISSUE TITLE
[Improvement] Some missing field/structure of vic3-tiger

### ISSUE THREAD

**Issue by dragon-archer** [amtep/ck3-tiger#78]
* Special event targets like `num_enemy_units` can be used like 'value triggers', while requiring input data scope inside parens
* `name` property allows block to implement dynamic naming (just like those in `desc`)
* Seems like `limited_to_frontier_colonization` in `law` is still not recognized by tiger
* `trigger_event` can also trigger on_actions, like `trigger_event = { on_action = XXX days = 1 }`
* Several properties in `subject_types`: `can_be_used_in_sway_offers` (boolean), `lower_autonomy_subject_type_alternatives` (array), `same_autonomy_subject_type_alternatives` (array), `higher_autonomy_subject_type_alternatives` (array), `ai_value` (value block), `can_target_with_transfer_wargoal` (boolean)
* `on_researched` in `technology`

**Comment by dragon-archer** [comment]
Some more items:
* `target_state` in `add_war_goal`
* `shadow_only` from `gfx\models\buildings\monuments\temple_of_poseidon\monument_templeofposeidon.asset` (dlc007_ap1)

**Comment by amtep** [comment]
> * Special event targets like `num_enemy_units` can be used like 'value triggers', while requiring input data scope inside parens
`num_enemy_units` was removed in 1.6 :) Do you have other examples? Tiger does support it but we may have missed some triggers where it applies.

>     * `name` property allows block to implement dynamic naming (just like those in `desc`)
Do you know where exactly? My experience with ck3 is that not all name fields can handle it.

>     * Seems like `limited_to_frontier_colonization` in `law` is still not recognized by tiger
Aha it was added in the wrong place.
 
>     * `trigger_event` can also trigger on_actions, like `trigger_event = { on_action = XXX days = 1 }`
Will add
 
>     * Several properties in `subject_types`: `can_be_used_in_sway_offers` (boolean), `lower_autonomy_subject_type_alternatives` (array), `same_autonomy_subject_type_alternatives` (array), `higher_autonomy_subject_type_alternatives` (array), `ai_value` (value block), `can_target_with_transfer_wargoal` (boolean)
Will do

>     * `on_researched` in `technology`
Will add

> `target_state` in `add_war_goal`
Will add

`shadow_only` from `gfx\models\buildings\monuments\temple_of_poseidon\monument_templeofposeidon.asset` (dlc007_ap1)
I don't have that dlc but I'll add it :) Can you say exactly where, like what is it a field of? Is it a boolean?

**Comment by dragon-archer** [comment]
> > * Special event targets like `num_enemy_units` can be used like 'value triggers', while requiring input data scope inside parens
> 
> `num_enemy_units` was removed in 1.6 :) Do you have other examples? Tiger does support it but we may have missed some triggers where it applies.

Ah, I still use 1.5.13, may I request a feature to run tiger under specific version content? That is, for each trigger/effect/event_target there'll be a supported version range (like the official wiki), while the version can be specified in vic3-tiger.conf or directly read from metadata/descriptor


> > ```
> > * `name` property allows block to implement dynamic naming (just like those in `desc`)
> > ```
> 
> 
> Do you know where exactly? My experience with ck3 is that not all name fields can handle it.

`common/scripted_buttons/brazilian_scripted_buttons.txt`

> > `shadow_only` from `gfx\models\buildings\monuments\temple_of_poseidon\monument_templeofposeidon.asset` (dlc007_ap1)
> 
> I don't have that dlc but I'll add it :) Can you say exactly where, like what is it a field of? Is it a boolean?

Yes, it's a boolean, I'll paste the whole file below
```
pdxmesh = {
	name = "monument_templeofposeidon_mesh"
	file = "monument_templeofposeidon.mesh"

	lod_percentages = {
		lod = { index = 1 percent = 16.0 }
		lod = { index = 2 percent = 8.0 }
		lod = { index = 3 percent = 4.0 }
	}

	meshsettings = {
		name = "templeofposeidon_hill_0Shape"
		index = 0
		shadow_only = yes
		shader = "standard"
		shader_file = "gfx/FX/pdxmesh.shader"
	}

	meshsettings = {
		name = "templeofposeidon_Shape0"
		index = 0
		texture_diffuse = "monument_templeofposeidon_diffuse.dds"
		texture_normal = "monument_templeofposeidon_normal.dds"
		texture_specular = "monument_templeofposeidon_properties.dds"
		texture = { file = "monument_templeofposeidon_tint.dds" index = 5 }
		shader = "standard"
		additional_shader_defines = { "TINT_COLOR" }
		shader_file = "gfx/FX/pdxmesh.shader"
	}

	meshsettings = {
		name = "templeofposeidon_Shape0"
		index = 1
		texture_diffuse = "monument_templeofposeidon_decal_diffuse.dds"
		texture_normal = "monument_templeofposeidon_decal_normal.dds"
		texture_specular = "monument_templeofposeidon_decal_properties.dds"
		texture = { file = "monument_templeofposeidon_tint.dds" index = 5 }
		shader = "standard"
		additional_shader_defines = { "TINT_COLOR" }
		shader_file = "gfx/FX/pdxmesh.shader"
	}

	meshsettings = {
		name = "templeofposeidon_Shape0"
		index = 2
		texture_diffuse = "monument_templeofposeidon_diffuse.dds"
		texture_normal = "monument_templeofposeidon_normal.dds"
		texture_specular = "monument_templeofposeidon_properties.dds"
		shader = "standard"
		additional_shader_defines = { "PDX_MESH_SNAP_VERTICES_TO_TERRAIN" "TINT_COLOR" }
		shader_file = "gfx/FX/pdxmesh.shader"
	}

	meshsettings = {
		name = "templeofposeidon_hill_1Shape"
		index = 0
		shadow_only = yes
		shader = "standard"
		shader_file = "gfx/FX/pdxmesh.shader"
	}

	meshsettings = {
		name = "templeofposeidon_Shape1"
		index = 0
		texture_diffuse = "monument_templeofposeidon_diffuse.dds"
		texture_normal = "monument_templeofposeidon_normal.dds"
		texture_specular = "monument_templeofposeidon_properties.dds"
		texture = { file = "monument_templeofposeidon_tint.dds" index = 5 }
		shader = "standard"
		additional_shader_defines = { "TINT_COLOR" }
		shader_file = "gfx/FX/pdxmesh.shader"
	}

	meshsettings = {
		name = "templeofposeidon_Shape1"
		index = 1
		texture_diffuse = "monument_templeofposeidon_diffuse.dds"
		texture_normal = "monument_templeofposeidon_normal.dds"
		texture_specular = "monument_templeofposeidon_properties.dds"
		shader = "decal_local"
		shader_file = "gfx/FX/pdxmesh_decal.shader"
		subpass = "LocalDecals"
	}

	meshsettings = {
		name = "templeofposeidon_Shape1"
		index = 2
		texture_diffuse = "monument_templeofposeidon_decal_diffuse.dds"
		texture_normal = "monument_templeofposeidon_decal_normal.dds"
		texture_specular = "monument_templeofposeidon_decal_properties.dds"
		shader = "standard"
		shader_file = "gfx/FX/pdxmesh.shader"
	}

	meshsettings = {
		name = "templeofposeidon_hill_2Shape"
		index = 0
		shadow_only = yes
		shader = "standard"
		shader_file = "gfx/FX/pdxmesh.shader"
	}

	meshsettings = {
		name = "templeofposeidon_Shape2"
		index = 0
		texture_diffuse = "monument_templeofposeidon_diffuse.dds"
		texture_normal = "monument_templeofposeidon_normal.dds"
		texture_specular = "monument_templeofposeidon_properties.dds"
		texture = { file = "monument_templeofposeidon_tint.dds" index = 5 }
		shader = "standard"
		additional_shader_defines = { "TINT_COLOR" }
		shader_file = "gfx/FX/pdxmesh.shader"
	}

	meshsettings = {
		name = "templeofposeidon_Shape2"
		index = 1
		texture_diffuse = "monument_templeofposeidon_diffuse.dds"
		texture_normal = "monument_templeofposeidon_normal.dds"
		texture_specular = "monument_templeofposeidon_properties.dds"
		shader = "decal_local"
		shader_file = "gfx/FX/pdxmesh_decal.shader"
		subpass = "LocalDecals"
	}

	meshsettings = {
		name = "templeofposeidon_Shape2"
		index = 2
		texture_diffuse = "monument_templeofposeidon_decal_diffuse.dds"
		texture_normal = "monument_templeofposeidon_decal_normal.dds"
		texture_specular = "monument_templeofposeidon_decal_properties.dds"
		shader = "standard"
		shader_file = "gfx/FX/pdxmesh.shader"
	}

	meshsettings = {
		name = "templeofposeidon_hill_3Shape"
		index = 0
		shadow_only = yes
		shader = "standard"
		shader_file = "gfx/FX/pdxmesh.shader"
	}

	meshsettings = {
		name = "templeofposeidon_Shape3"
		index = 0
		texture_diffuse = "monument_templeofposeidon_diffuse.dds"
		texture_normal = "monument_templeofposeidon_normal.dds"
		texture_specular = "monument_templeofposeidon_properties.dds"
		texture = { file = "monument_templeofposeidon_tint.dds" index = 5 }
		shader = "standard"
		additional_shader_defines = { "TINT_COLOR" }
		shader_file = "gfx/FX/pdxmesh.shader"
	}
}

entity = {
	name = "monument_templeofposeidon_entity"
	pdxmesh = "monument_templeofposeidon_mesh"
	locator = { name = "loc_lights" position = { 0.0 0.0 0.0 }}
	default_state = "idle"
		
	state = { 
		name = "idle" 
		state_time = 5

		event = { time = 0 node = "loc_lights" particle = "infrastructure/poseidon_mist" trigger_once = yes}
	}
}
```

**Comment by dragon-archer** [comment]
Some more things:
* `startangle` and `endangle` in `gui/politics_panel_types.gui` under `piechart`

**Comment by amtep** [comment]
>     * property `startangle` and `endangle` in `gui/politics_panel_types.gui` under `piechart`

Will add

>     * data function `SelectMapModeByKey('mm_XXX')` and `UnforceMapMode` in `gui/states_panel.gui`

These were also removed in 1.6. I think UnforceMapMode was replaced with UnlockMapMode.

**Comment by amtep** [comment]
> may I request a feature to run tiger under specific version content?

I'm afraid that would be a too large maintenance commitment. Every part of tiger would be affected; not just all the tables, but also the schemas for validating all the items in common/.

Your best option would be to hold back the ck3-tiger version to one that matches the vic3 release you are using. You won't get improvements that way, but you'll get fewer false positives based on the release differences.

I might change my mind on this if you (and other users) plan to stay on Vic3 1.5 long term. Which I guess would mean that 1.6 was pretty bad :)

**Comment by dragon-archer** [comment]
> These were also removed in 1.6. I think UnforceMapMode was replaced with UnlockMapMode.

So perhaps we can treat these removed things like those renamed ones (like `culture_can_have_mass_migration_to`) to produce a special `removed` error (maybe together with a removed version) instead of just remove them from the table?

> Your best option would be to hold back the ck3-tiger version to one that matches the vic3 release you are using. You won't get improvements that way, but you'll get fewer false positives based on the release differences.

Currently I simply filter out these false postives, so if all these removed old things can be marked with a version mark, then it world be quite easier to filter out them.

> I'm afraid that would be a too large maintenance commitment. Every part of tiger would be affected; not just all the tables, but also the schemas for validating all the items in common/.

Thus perhaps we can remain those recently removed items (i.e. those removed from the 1 or 2 most recent minor version) in table with a suggestion?

> I might change my mind on this if you (and other users) plan to stay on Vic3 1.5 long term. Which I guess would mean that 1.6 was pretty bad :)

Actually I have no plan to stick to any version for a long term, I'm just waiting for Paradox to publish more hot fixes to make the version stable. You know, Paradox always mess up things...

**Comment by amtep** [comment]
Closing because all the specific items listed have been fixed in main.

> So perhaps we can treat these removed things like those renamed ones (like `culture_can_have_mass_migration_to`) to produce a special `removed` error (maybe together with a removed version) instead of just remove them from the table?

There are some technical difficulties with implementing this so I've moved it to the IDEAS wiki page. https://github.com/amtep/ck3-tiger/wiki/Ideas

(To be specific: the possibility that a function is removed for only one datatype but remains for others, and the removal of entire datatypes. Both are fairly common in updates.)
